File: db4on/reference/db4o-reference-chapters/src/com/db4odoc/f1/uuids/UUIDExample.cs
/* Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com */
using System;
using System.IO;
using com.db4o;
using com.db4o.ext;
using com.db4o.query;
using com.db4o.foundation;

namespace com.db4odoc.f1.uuids
{
	public class UUIDExample
	{
		public readonly static string YapFileName = "formula1.yap";

		public static void Main(String[] args) 
		{
			TestChangeIdentity();
			SetObjects();
			TestGenerateUUID();
		}
		// end Main
	
		private static string PrintSignature(byte[] Signature)
		{
			String str="";
			for (int i = 0; i < Signature.Length; i++) 
			{
				str = str+Signature[i];
			}
			return str;
		}
		// end PrintSignature

		public static void TestChangeIdentity()
		{
			File.Delete(YapFileName);
			ObjectContainer oc = Db4o.OpenFile(YapFileName);
			Db4oDatabase db;
			byte[] oldSignature;
			byte[] newSignature;
			try 
			{
				db = oc.Ext().Identity();
				oldSignature = db.GetSignature();
				Console.WriteLine("oldSignature: " + PrintSignature(oldSignature));	
				((YapFile)oc).GenerateNewIdentity();
			} 
			finally 
			{
				oc.Close();
			}        
			oc = Db4o.OpenFile(YapFileName);
			try 
			{
				db = oc.Ext().Identity();
				newSignature = db.GetSignature();
				Console.WriteLine("newSignature: " + PrintSignature(newSignature));
			} 
			finally 
			{
				oc.Close();
			}
        
			bool same = true;
        
			for (int i = 0; i < oldSignature.Length; i++) 
			{
				if(oldSignature[i] != newSignature[i])
				{
					same =false;
				}
			}
        
			if (same)
			{
				Console.WriteLine("Database signatures are identical");
			} 
			else 
			{
				Console.WriteLine("Database signatures are different");
			}
		}
		// end TestChangeIdentity

		public static void SetObjects()
		{
			Db4o.Configure().ObjectClass(typeof(Pilot)).GenerateUUIDs(true);
			File.Delete(YapFileName);
			ObjectContainer oc = Db4o.OpenFile(YapFileName);
			try 
			{
				Car car = new Car("BMW", new Pilot("Rubens Barrichello"));
				oc.Set(car);
			} 
			finally 
			{
				oc.Close();
			}
		}
		// end SetObjects

		public static void TestGenerateUUID()
		{
			ObjectContainer oc = Db4o.OpenFile(YapFileName);
			try 
			{
				Query query = oc.Query();
				query.Constrain(typeof(Car));
				ObjectSet result = query.Execute();
				Car car = (Car)result[0];
				ObjectInfo carInfo = oc.Ext().GetObjectInfo(car);
				Db4oUUID carUUID = carInfo.GetUUID();
				Console.WriteLine("UUID for Car class are not generated:");
				Console.WriteLine("Car UUID: " + carUUID);
			
				Pilot pilot = car.Pilot;
				ObjectInfo pilotInfo = oc.Ext().GetObjectInfo(pilot);
				Db4oUUID pilotUUID = pilotInfo.GetUUID();
				Console.WriteLine("UUID for Car class are not generated:");
				Console.WriteLine("Pilot UUID: " + pilotUUID);
				Console.WriteLine("long part: " + pilotUUID.GetLongPart() +"; signature: " + PrintSignature(pilotUUID.GetSignaturePart()));
				long ms = TimeStampIdGenerator.IdToMilliseconds(pilotUUID.GetLongPart());
				Console.WriteLine("Pilot object was created: " + (new DateTime(1970,1,1)).AddMilliseconds(ms).ToString());
				Pilot pilotReturned = (Pilot)oc.Ext().GetByUUID(pilotUUID);
				Console.WriteLine("Pilot from UUID: " + pilotReturned);	
			} 
			finally 
			{
				oc.Close();
			}        
		}
		// end TestGenerateUUID
	}
}

File: db4on/reference/vb/db4o-reference-chapters/src/com/db4odoc/f1/uuids/UUIDExample.vb
' Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com 
Imports System
Imports System.IO
Imports com.db4o
Imports com.db4o.query
Imports com.db4o.foundation
Imports com.db4o.ext


Namespace com.db4odoc.f1.uuids
    Public Class UUIDExample
        Inherits Util

        Public Shared Sub main(ByVal args() As String)
            TestChangeIdentity()
            SetObjects()
            TestGenerateUUID()
        End Sub

        Private Shared Function PrintSignature(ByVal Signature() As Byte) As String
            Dim str As String = ""
            Dim i As Integer
            For i = 0 To Signature.Length - 1 Step i + 1
                str = str + Signature(i).ToString()
            Next
            Return str
        End Function

        Public Shared Sub TestChangeIdentity()

            File.Delete(Util.YapFileName)
            Dim oc As ObjectContainer = Db4o.OpenFile(Util.YapFileName)
            Dim db As Db4oDatabase
            Dim oldSignature() As Byte
            Dim NewSignature() As Byte
            Try
                db = oc.Ext().Identity()
                oldSignature = db.GetSignature()
                Console.WriteLine("oldSignature: " + PrintSignature(oldSignature))
                Dim yf As YapFile = DirectCast(oc, YapFile)
                yf.GenerateNewIdentity()
            Finally
                oc.Close()
            End Try
            oc = Db4o.OpenFile(Util.YapFileName)
            Try
                db = oc.Ext().Identity()
                NewSignature = db.GetSignature()
                Console.WriteLine("newSignature: " + PrintSignature(NewSignature))
            Finally
                oc.Close()
            End Try

            Dim same As Boolean = True

            Dim i As Integer
            For i = 0 To oldSignature.Length - 1 Step i + 1
                If oldSignature(i) <> NewSignature(i) Then

                    same = False
                End If
            Next

            If (same) Then
                Console.WriteLine("Database signatures are identical")
            Else
                Console.WriteLine("Database signatures are different")
            End If
        End Sub

        Public Shared Sub SetObjects()
            Db4o.Configure().ObjectClass(GetType(Pilot)).GenerateUUIDs(True)
            File.Delete(Util.YapFileName)
            Dim oc As ObjectContainer = Db4o.OpenFile(Util.YapFileName)
            Try
                Dim car As Car = New Car("BMW", New Pilot("Rubens Barrichello"))
                oc.Set(car)
            Finally
                oc.Close()
            End Try
        End Sub

        Public Shared Sub TestGenerateUUID()
            Dim oc As ObjectContainer = Db4o.OpenFile(Util.YapFileName)
            Try
                Dim query As Query = oc.Query()
                query.Constrain(GetType(car))
                Dim result As ObjectSet = query.Execute()
                Dim car As Car = CType(result(0), Car)
                Dim carInfo As ObjectInfo = oc.Ext().GetObjectInfo(car)
                Dim carUUID As Db4oUUID = carInfo.GetUUID()
                Console.WriteLine("UUID for Car class are not generated:")
                If carUUID Is Nothing Then
                    Console.WriteLine("Car UUID: null")
                Else
                    Console.WriteLine("Car UUID: " + carUUID.ToString())
                End If


                Dim pilot As Pilot = car.Pilot
                Dim pilotInfo As ObjectInfo = oc.Ext().GetObjectInfo(pilot)
                Dim pilotUUID As Db4oUUID = pilotInfo.GetUUID()
                Console.WriteLine("UUID for Car class are not generated:")
                If pilotUUID Is Nothing Then
                    Console.WriteLine("Pilot UUID: null")
                Else
                    Console.WriteLine("Pilot UUID: " + pilotUUID.ToString())
                End If

                Console.WriteLine("long part: " + pilotUUID.GetLongPart().ToString() + "; signature: " + PrintSignature(pilotUUID.GetSignaturePart()))
                Dim ms As Long = TimeStampIdGenerator.IdToMilliseconds(pilotUUID.GetLongPart())
                Console.WriteLine("Pilot object was created: " + (New DateTime(1970, 1, 1)).AddMilliseconds(ms).ToString())
                Dim pilotReturned As Pilot = CType(oc.Ext().GetByUUID(pilotUUID), Pilot)
                Console.WriteLine("Pilot from UUID: " + pilotReturned.ToString())
            Finally
                oc.Close()
            End Try
        End Sub

    End Class
End Namespace